File: AC/AC_II/CodigoDeHamming/hamming.c
#include <stdio.h>
#include <stdbool.h>

#define length(array) (int)(sizeof(array) / sizeof(array[0]))

int hammingPos(int n, int bits[]) {
	int errorPos = 0;

	for (int i = 0; i < n; i++) {
		if (!bits[i]) errorPos ^= i;
	}

	return errorPos;
}

int hammingError(int n, int bits[]) {
	// Conta a quantidade de UM's
	int onesCount = 0;
	for (int i = 0; i < n; i++) {
		if (bits[i] == 1) onesCount++;
	}

	// Verifica se o valor da posição 0 condiz com a quantidade geral de UM's
	bool overallParity = bits[0] == 0 && onesCount % 2 == 0 ? true : false;

	// Verifica se há erro de paridade em algum dos "setores"
	bool sectorParity = hammingPos(n, bits) == 0;

	int result = 0;

	if (overallParity && !sectorParity) result = 2;
	else if (!overallParity || !sectorParity) result = 1;

	return result;
}

void main() {
	int bits[] = {
		0, 0, 0, 1,
		1, 1, 0, 1,
		1, 1, 1, 0,
		0, 0, 1, 0
	};

	switch (hammingError(length(bits), bits)) {
		case 2:
			printf("Mais que dois erros, abortar!\n");
		break;
		case 1:
			int pos = hammingPos(length(bits), bits);
			printf("Apenas um erro detectado na posição: %d.\n", pos);
			bits[pos] = !bits[pos];
		break;
		case 0:
			printf("Nenhum erro detectado.\n");
		break;
	}
}

File: AC/AC_II/CodigoDeHamming/hamming.py
from functools import reduce

import operator as op

hammingPos = lambda bits: reduce(op.xor, [i for (i, bit) in enumerate(bits) if bit])

def hammingError(bits):
	# Conta a quantidade de UM's
	onesCount = reduce(lambda x, y: x + y, [bit for bit in bits if bit == 1])

	# Verifica se o valor da posição 0 condiz com a quantidade geral de UM's
	overallParity = True if bits[0] == 0 and onesCount % 2 == 0 else False

	# Verifica se há erro de paridade em algum dos setores
	sectorParity = hammingPos(bits) == 0

	result = 0
	if overallParity and not sectorParity: result = 2
	elif not overallParity or not sectorParity: result = 1

	return result

def main():
	bits = [
		0, 0, 0, 1,
		1, 1, 0, 1,
		1, 1, 1, 0,
		0, 0, 1, 0
	]

	match hammingError(bits):
		case 2:
			print("Mais que dois erros, abortar!")
		case 1:
			print("Apenas um erro detectado na posição: %s." %(hammingPos(bits)))
		case 0:
			print("Nenhum erro detectado.")

main()

def foo():
	list = [5,6,7,8,9]
	list2 = [(i, b) for (i, b) in enumerate(list)]
	print(list)
	print(list2)

	reduced = reduce(lambda x, y: x + y, list)
	print(reduced)

# foo()